group-coordinator/src/main/java/org/apache/kafka/coordinator/group/classic/ClassicGroup.java

```diff
@@ -1300,6 +1341,68 @@ public Map<String, byte[]> groupAssignment() {
         ));
     }
 
+    /**
+     * Convert the current classic group to a consumer group.
+     * Add the records for the conversion.
+     *
+     * @param consumerGroup     The converted consumer group.
+     * @param records           The list to which the new records are added.
+     *
+     * @throws GroupIdNotFoundException if any of the group's member doesn't support the consumer protocol.
+     */
+    public void convertToConsumerGroup(
+        ConsumerGroup consumerGroup,
+        List<Record> records,
+        TopicsImage topicsImage
+    ) throws GroupIdNotFoundException {
+        consumerGroup.setGroupEpoch(generationId);
+        consumerGroup.setTargetAssignmentEpoch(generationId);
+
+        records.add(RecordHelpers.newGroupEpochRecord(groupId(), generationId));
+        // SubscriptionMetadata will be computed in the following consumerGroupHeartbeat
+        records.add(RecordHelpers.newGroupSubscriptionMetadataRecord(groupId(), Collections.emptyMap()));
+        records.add(RecordHelpers.newTargetAssignmentEpochRecord(groupId(), generationId));
+        
+        members.forEach((memberId, member) -> {
+            ConsumerPartitionAssignor.Assignment assignment = ConsumerProtocol.deserializeAssignment(ByteBuffer.wrap(member.assignment()));
+            Map<Uuid, Set<Integer>> partitions = topicPartitionMapFromList(assignment.partitions(), topicsImage);
+            ConsumerPartitionAssignor.Subscription subscription = ConsumerProtocol.deserializeSubscription(ByteBuffer.wrap(member.metadata()));
+
+            ConsumerGroupMember newMember = new ConsumerGroupMember.Builder(memberId)
+                .setMemberEpoch(generationId)
+                .setPreviousMemberEpoch(generationId)
+                .setInstanceId(member.groupInstanceId().orElse(null))
+                .setRackId(subscription.rackId().orElse(null))
+                .setRebalanceTimeoutMs(member.rebalanceTimeoutMs())
+                .setClientId(member.clientId())
+                .setClientHost(member.clientHost())
+                .setSubscribedTopicNames(subscription.topics())
+                .setAssignedPartitions(partitions)
+                .build();
+            consumerGroup.updateMember(newMember);
+
+            records.add(RecordHelpers.newMemberSubscriptionRecord(groupId(), newMember));
+            records.add(RecordHelpers.newCurrentAssignmentRecord(groupId(), newMember));
+            records.add(RecordHelpers.newTargetAssignmentRecord(groupId(), memberId, partitions));
+        });
+    }
+
+    private static Map<Uuid, Set<Integer>> topicPartitionMapFromList(
+        List<TopicPartition> partitions,
+        TopicsImage topicsImage
+    ) {
+        Map<Uuid, Set<Integer>> topicPartitionMap = new HashMap<>();
+        partitions.forEach(topicPartition -> {
+            TopicImage topicImage = topicsImage.getTopic(topicPartition.topic());
+            // TODO: add a log if topic image is null?
```

Comment:
I suppose that we could just ignore them. If the image is null, the topic does not exist so it must be removed anyway.